<a href="https://colab.research.google.com/github/abhishekmishra-bareilly/Machine-learning/blob/main/Diabetes_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Diabetes Prediction

**Import Dependancy**

In [1]:
# Import Dependancy
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

**Import dataset**

In [2]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Machine Learning/Learn from outside/type 1/Copy of diabetes.csv')

In [3]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
# Shape of the data
data.shape

(768, 9)

**Diabetes dataset :-**


*   In the data we have **768 rows** and **9 columns**.
*   In this data column **"Outcome"** is the **lable**.


---





In [6]:
# describe the data
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [7]:
# Check the label data distribution
data['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64



*   So 0 have 500 rows and 1 have 268 so we can say the data is not equally distributed
*   0 --> Not Diabetec
* 1--> Diabetes


---





**Check the mean values for "outcome" column**

In [8]:
# Check the mean values for outcome column
data.xgroupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164



---



**Separating the data as x and labels y**

In [16]:
# Separating the data as x and labels y
x = data.drop(columns= 'Outcome', axis = 1)
y = data['Outcome']

In [17]:
print(x)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [18]:
print(y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64




---



**Standariz the data**

In [20]:
scaler = StandardScaler()

In [21]:
scaler.fit(x)

StandardScaler()

In [22]:
scaler_x = scaler.transform(x)

In [24]:
x = scaler_x
x

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])



---



**Train Test Split**

In [31]:
# Train Test Split the data
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2, stratify=y, random_state=2)

In [32]:
# See the shape of the train data
print(x.shape, x_train.shape, x_test.shape)

(768, 8) (614, 8) (154, 8)




---



**Training the Model using support vector Machine Classifier**

In [33]:
# Training the Model using support vector Machine Classifier
classifier = svm.SVC(kernel='linear')

In [34]:
#training the support vector Machine Classifier
classifier.fit(x_train, y_train)

SVC(kernel='linear')



---



**Model Evaluation**

**Accuracy Score on training data**

In [40]:
# Accuracy Score on training data
x_train_pred = classifier.predict(x_train)
training_data_Accuracy = accuracy_score(x_train_pred,y_train)
print('Accuracy Score on training data :-',training_data_Accuracy)

Accuracy Score on training data :- 0.7866449511400652


**Accuracy Score on test data**

In [41]:
# Accuracy Score on test data
x_test_pred = classifier.predict(x_test)
test_data_Accuracy = accuracy_score(x_test_pred,y_test)
print('Accuracy Score on test data :-',test_data_Accuracy)

Accuracy Score on test data :- 0.7727272727272727




*   so our model is not **over fitted**.


---






**Making a Predictive System**

In [44]:
# Making a Predictive System
input_data = (5,166,72,19,175,25.8,0.587,51)

# changing the input_data to numpy array
input_data_as_np_array =np.asarray(input_data)

# reshape the array as we are predicting for one instance
data_reshape = input_data_as_np_array.reshape(1,-1)

# standardize the input data
std_data = scaler.transform(data_reshape)
print(std_data)

# prediction of the data
prediction = classifier.predict(std_data)
print(prediction)

# Making better predictation 
if (prediction[0] == 0):
  print('The person is not diabetic')
else:
  print('The person is diabetic')

[[ 0.3429808   1.41167241  0.14964075 -0.09637905  0.82661621 -0.78595734
   0.34768723  1.51108316]]
[1]
The person is diabetic


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  "X does not have valid feature names, but"
